storage/remote/queue_manager.go

```diff
@@ -873,57 +982,78 @@ func (s *shards) sendSamples(ctx context.Context, samples []prompb.TimeSeries, b
 
 // sendSamples to the remote storage with backoff for recoverable errors.
 func (s *shards) sendSamplesWithBackoff(ctx context.Context, samples []prompb.TimeSeries, buf *[]byte) error {
-	backoff := s.qm.cfg.MinBackoff
-	req, highest, err := buildWriteRequest(samples, *buf)
+	// Build the WriteRequest with no metadata.
+	req, highest, err := buildWriteRequest(samples, nil, *buf)
 	*buf = req
 	if err != nil {
 		// Failing to build the write request is non-recoverable, since it will
 		// only error if marshaling the proto to bytes fails.
 		return err
 	}
 
+	reportFailureFunc := func() {
+		s.qm.retriedSamplesTotal.Add(float64(len(samples)))
+	}
+	elapsed, err := sendWriteRequestWithBackoff(ctx, s.qm.cfg, s.qm.client, s.qm.logger, req, reportFailureFunc)
+	if err != context.Canceled {
+		s.qm.sentBatchDuration.Observe(elapsed)
+	}
+	if err != nil {
+		return err
+	}
+	s.qm.succeededSamplesTotal.Add(float64(len(samples)))
+	s.qm.bytesSent.Add(float64(len(*buf)))
+	s.qm.highestSentTimestampMetric.Set(float64(highest / 1000))
+	atomic.StoreInt64(&s.qm.lastSendTimestamp, time.Now().Unix())
+	return nil
+}
+
+func sendWriteRequestWithBackoff(ctx context.Context, cfg config.QueueConfig, s func() StorageClient, l log.Logger, req []byte, reportFailureFunc func()) (elapsed float64, err error) {
```

Comment:
The conflict I had came from this function - before, it was receiving a `StorageClient` directly. Now, it takes a function which is `client()` that returns the `StorageClient` but with a lock.